### PR TITLE
atommapping implementation

### LINKED ISSUE #6: Implement AtomMapping container
Should be a read only simple data class holding enough information for mapping two sets of ints (atom indices) to each other...

```python
class AtomMapping:
    AtoB : dict[int, int]
    BtoA : dict[int, int]
```

### PR DESCRIPTION
**Pull request by richardjgowers**
Can probably leave it this simple, makes it easy for others to only use the AtomMappers.

Resolves #6.

### CODE REVIEW

**Comment by IAlibay** [comment]
Just having a quick glance - is there a reason for having a Perses specific method here? Ideally we should try to build this in a way that we don't have to change this class every time someone comes up with a new mapper right?

File: openfe/setup/atommapping.py [hunk 1]
```diff
@@ -1,8 +1,41 @@
...
-        return None
 
+    The size of molecule A/B is given by the length of the AtoB/BtoA dictionary
+    """
+    molA: RDKitMol
+    molB: RDKitMol
+    AtoB: Dict[int, Union[int, None]]
+    BtoA: Dict[int, Union[int, None]]
+
+    def __init__(self, molA: RDKitMol, molB: RDKitMol):
```
Comment (dwhswenson) [inline openfe/setup/atommapping.py:28]: Any reason not to include the dicts in the init sig? I think it's easier for a user to write:

```python
return AtomMapping(molA, molB, myAtoB, myBtoA)
```

than

```python
mapping = AtomMapping(molA, molB)
mapping.AtoB = myAtoB
mapping.BtoA = myBtoA
return mapping
```

If you're worried about users mixing up order, make them keyword required. If not, maybe the whole thing is `@dataclass`?

**Review by dwhswenson — CHANGES_REQUESTED** [review]
A couple small changes. (And you may get two versions of one of my comments, since you updated the PR, fixing one of the issues, when I was reviewing it!)

And I see you just updated the PR again as I was typing this. 😂 Hope my review is still relevant!

File: openfe/setup/atommapping.py [hunk 2]
```diff
@@ -1,8 +1,27 @@
...
+      maps the index of an atom in either molecule **A** or **B** to the other.
+      If this atom has no corresponding atom, None is returned.
 
+
+    The size of molecule A/B is given by the length of the AtoB/BtoA dictionary
+    """
+    AtoB: Dict[int, Union[int, None]]
+    BtoA: Dict[int, Union[int, None]]
+
+    def __init__(self):
```
[the remark this answers is not in the document]
Comment (richardjgowers) [reply openfe/setup/atommapping.py]: Yeah I realised point 2 a little later than you

File: openfe/setup/atommapping.py [hunk 3]
```diff
@@ -1,8 +1,34 @@
...
+    mol1, mol2 : rdkit.Mol
+      the two Molecules in the mapping
+    mol1_to_mol2, mol2_to_mol1 : dict
+      maps the index of an atom in either molecule **A** or **B** to the other.
+      If this atom has no corresponding atom, None is returned.
 
+    """
+    mol1: RDKitMol
+    mol2: RDKitMol
+    mol1_to_mol2: Dict[int, Union[int, None]]
```
Comment (dwhswenson) [inline openfe/setup/atommapping.py:25]: Are values here still `Union[int, None]` or are they now always `int`?

Comment (IAlibay) [reply openfe/setup/atommapping.py:25]: they _should_ just be `int`

Comment (dwhswenson) [inline openfe/setup/atommapping.py:25]:
```suggestion
    mol1_to_mol2: Dict[int, int]
```

Suggesting and accepting suggestion at suggestion of @richardjgowers

**Review by IAlibay — APPROVED** [review]

**Review by dwhswenson — APPROVED** [review]
LGTM; will merge. Just commenting on a couple of things that might change later, but don't need to change for this PR:

1. I'd love to see better names for fixtures `simple_mapping` and `other_mapping`, and then to move them to `conftest.py` and allow them to be reused by other tests.

2. The name `mol1_to_mol2` might change because we might want that name for the lookups in [#8](https://github.com/OpenFreeEnergy/openfe/issues/8) (along with `mol2_to_mol1`).

**Pull request merged**